pyscf/mcscf/addons.py

```diff
@@ -307,6 +307,46 @@ def sort_mo_by_irrep(casscf, mo_coeff, cas_irrep_nocc,
                              cas_irrep_ncore, s, base=0)
     return sort_mo(casscf, mo_coeff, caslst, base=0)
 
+def make_natural_orbitals (method_obj):
+    """Make natural orbitals from a general PySCF method object. 
+    See Eqn. (1) in Keller et al. [DOI:10.1063/1.4922352] for details.
+
+
+    Args:
+        method_obj : Any PySCF method that has the function `make_rdm1` with kwarg 
+            `ao_repr`. This object can be a restricted OR unrestricted method.
+
+    Returns:
+        noons : A 1-D array of the natural orbital occupations (NOONs). 
+        
+        natorbs : A set of natural orbitals made from method_obj.
+    """
+    mf = method_obj._scf
+    rdm1 = method_obj.make_rdm1(ao_repr=True)
+    S = mf.get_ovlp()
+
+    # Slight difference for restricted vs. unrestriced case
+    if isinstance(rdm1, tuple):
+        Dm = rdm1[0]+rdm1[1]
+    elif isinstance(rdm1, numpy.ndarray):
+        Dm = rdm1
+    else:
+        raise ValueError("\n\tThe rdm1 generated by method_obj.make_rdm1() was a {}.".format(type(rdm1))
+            + 
+            "\n\tThis type is not supported, please select a different method and/or open an issue at https://github.com/pyscf/pyscf/issues"
+        )
+
+    # Diagonalize the DM in AO (using Eqn. (1) referenced above)
+    A = reduce(numpy.dot, (S, Dm, S))
+    w, v = scipy.linalg.eigh(A, b = S)
```

Comment:
You shouldn't use scipy here; you should use the method's `eig` which may have been overwritten

Reply:
This implementation is not safe in linearly dependent basis sets.

A better alternative would be to diagonalize in the basis of the molecular orbitals, which form an orthonormal set.